feather/server/src/main.rs

```diff
@@ -58,21 +69,111 @@ fn init_systems(game: &mut Game, server: Server) {
+
+        game.ecs.spawn((
+            WorldName::new(world.to_string()),
+            WorldPath::new(PathBuf::from(format!("worlds/{}", world))),
+            Dimensions::default(),
+        ));
+    }
+}
+
+fn init_dimensions(game: &mut Game, config: &Config) {
+    let biomes = game.resources.get::<Arc<BiomeList>>().unwrap();
+    for namespace in std::fs::read_dir("worldgen")
+        .expect("There's no worldgen/ folder. Try removing generated/ and re-running feather")
+    {
+        let namespace_dir = namespace.unwrap().path();
+        for file in std::fs::read_dir(namespace_dir.join("dimension")).unwrap() {
+            let mut dimension_info: DimensionInfo = serde_json::from_str(
+                &std::fs::read_to_string(file.as_ref().unwrap().path()).unwrap(),
+            )
+            .unwrap();
+
+            dimension_info.info = serde_json::from_str(
+                &std::fs::read_to_string(format!(
+                    "worldgen/{}/dimension_type/{}.json",
+                    dimension_info.r#type.split_once(':').unwrap().0,
+                    dimension_info.r#type.split_once(':').unwrap().1
+                ))
+                .unwrap(),
+            )
+            .unwrap();
+
+            for (_, (world_name, world_path, dimensions)) in game
+                .ecs
+                .query::<(&WorldName, &WorldPath, &mut Dimensions)>()
+                .iter()
+            {
+                if !dimensions
+                    .iter()
+                    .any(|dim| dim.info().r#type == dimension_info.r#type)
+                {
+                    let generator: Arc<dyn WorldGenerator> = match &config.worlds.generator[..] {
+                        "flat" => Arc::new(SuperflatWorldGenerator::new(
+                            SuperflatGeneratorOptions::default(),
+                        )),
+                        other => {
+                            log::error!("Invalid generator specified in config.toml: {}", other);
+                            std::process::exit(1);
+                        }
+                    };
+                    let is_flat = config.worlds.generator == "flat";
+
+                    log::info!(
+                        "Adding dimension `{}` to world `{}`",
+                        dimension_info.r#type,
+                        **world_name
+                    );
+                    dimensions.add(Dimension::new(
+                        dimension_info.clone(),
+                        generator,
+                        world_path
+                            .join("dimensions")
+                            .join(dimension_info.r#type.split_once(':').unwrap().0)
+                            .join(dimension_info.r#type.split_once(':').unwrap().1),
+                        false,
+                        is_flat,
+                        Arc::clone(&*biomes),
+                    ));
+                }
+            }
+        }
+    }
+}
+
+fn init_biomes(game: &mut Game) {
```

Comment:
This can only panic on non-utf8 file names, files not ending with .json, and invalid biome .json files, I think that failing fast on startup is better than making a user think "why doesn't my biome work?" because of a trailing comma.

Reply:
non-utf8 thing is just sad but whatever,
invalid biomes should ideally not panic if the biome is never used but also passable,
files not ending with .json immediately stops working on mac. you can rely on the OS to insert a .DS_Store file anywhere you look. also why not have texts or images or any other crap hidden in datapacks, who does it hurt?

P.S, yes, this is my first appearance on feather gh

Reply:
You're right, it should ignore .DS_Store, but images or text files shouldn't really exist in biomes/. Also, worldgen/ is not a datapack, though it would be great if we extract the full vanilla datapack when we implement the datapack support, so it's rather a temporary solution.

Reply:
Feather (or any non-prototype rust code for that matter) should only panic on programming errors, never on I/O errors, which can and will definitely occur in normal usage. `init_game` already returns `anyhow::Result` so you don't even have to make big changes through a lot of code to implement proper error handling. Just pass the error up the chain by using `?` and letting `init_biomes` return `anyhow::Result<()>`. This also applies for `init_dimensions` and anywhere else (I have not reviewed the PR yet).

Reply:
Nice work, looks good